### ISSUE TITLE
Support for ternary operator?

### ISSUE THREAD

**Issue by ThrDev** [lunet-io/scriban#34]
Is there something planned for a ternary operator on boolean expressions? For example, right now I'm doing something like this:
{{ if expr }} 
{{ function1 }}
{{ else }}
{{ function2 }}
{{ end }}

It would be really nice to be able to do:
{{ if expr ? function1 : function2 }} or something to that effect. 

Even something such as:
{{ if expr function1 function2 }}

**Comment by xoofx** [comment]
I have hesitated to introduce ternary operator, as they usually complicate the parsing.
From your proposal, not sure that they can really work with the language without introducing parsing conflicts.

- `{{ if expr ? function1 : function2 }}`: Conflicts with named parameters `name1 : x` passed to a function. function1 in that case could have named parameters and it would be annoying to workaround it, as you would have to use paranthesis to avoid parsing errors `(function1)`...
- `{{ if expr function1 function2 }}` is more conflicting, spaces are used to separate function arguments, so you can't really use them to separate function1 from function2. in this case, function1 and function2 would be considered as function arguments of expr.

As a workaround you could use `{{if expr; function1; else; function 2; end}}` if you want something on the same line. It is still not an expression and quite verbose, that the closest syntax for now.

**Comment by pebezo** [comment]
`Conflicts with named parameters name1 : x` 

My first reaction was ... wait, we don't have named params, but then I saw your post:

https://github.com/lunet-io/scriban/issues/28

Wow, nice job!! 👏

**Comment by ThrDev** [comment]
The last line for the workaround is fine, I won't be able to use that in some places though since I would have to concatenate 2 function outputs together with a " " space separator.

I appreciate the help!

**Comment by ebekker** [comment]
An option that I believe would be compatible with the parsing structure of the language would be something like this:

```
{{ iff test-expr; true-expr; false-expr }}
```

In this case you have another keyword that signals the conditional logic about to come and their are exactly 3 argument _single_ expressions to this keyword.  They are separated by semicolons in keeping with the existing rules of the language.  (In this example I chose `iff` because it corresponds to the mathematical *if-and-only-if* construct, but any short keyword would do.)

(Alternatively you could separate the expressions on multiple lines and omit the semicolons but that would conflict with the intended brevity of this notation.) 

Example:

```
<td class="{{ iff is_odd; 'odd-row'; 'even-row' }}">
```

Just seems a bit more concise and clean.  The key aspects are the omission of the `else` and `end` keywords, and required single-expression components.

**Comment by jods4** [comment]
Replaced some T4 templates with Scriban today and it's great. ❤️ 

I'm gonna say that emitting expressions based on a condition is fairly common and the `{{ if x; y; else; z; end }}` is really verbose for something that simple / common.

I wish we could have the common ternary expression `x ? y : z`, or at least a keyword `iif` that behaves like a function: `{{ iif x y z }}`.

BTW `iif` is borrowed from VB, it stands for _Immediate If_ and unlike the real function `If(cond, then, else)` it's a keyword that doesn't evaluate the branch that doesn't match the condition; which of course is key when your condition is something like `iif a != null (compute a.name) 'fallback'`

**Comment by xoofx** [comment]
I agree, common ternary expression is much less verbose then if/else, and it turns out this was not much trouble to implement.

It is in the commit 8e74689 and is part of Scriban 3.0 via the PR #230. It should be soon available in the NuGet package `3.0.0-alpha.6`

**Comment by jods4** [comment]
Thanks! 
Scriban is really great. 🙇

**Comment by xoofx** [comment]
Actually, there was a bug with operator precedence and conditional operator that should be fixed in `3.0.0-alpha.7`

**Comment by jods4** [comment]
@xoofx thanks for the heads up, I'm already using alpha.6 in my code and it works wonders, much nicer templates.

It's only very simple ternary expressions, so I did not notice the precedence problem but I'll upgrade anyway 👍
